### ISSUE TITLE
error in mod_http_upload_quota leads to shutdown

### ISSUE THREAD

**Issue by joudinet** [processone/ejabberd#2810]
I've modified mod_http_upload{,_quota} configurations to delete uploaded files after 30 days instead of when users are unregistered. See the diff of my changes:
```
--- a/ejabberd.yml
+++ b/ejabberd.yml
@@ -164,6 +164,9 @@ modules:
     access: all
     docroot: 'UPLOAD_ROOT'
     put_url: 'PUT_URL'
+    rm_on_unregister: false
+  mod_http_upload_quota:
+    max_days: 30
   mod_last: {}
   mod_mam:
     ## Mnesia is limited to 2GB, better to use an SQL backend
```

After restarting ejabberd, everything seems to work perfectly but, few hours later, ejabberd stops with the following error:
```
# cat /var/log/ejabberd/error.log 
2019-03-01 15:32:42.527 [error] <0.2021.0>@mod_http_upload_quota:handle_info:214 Got unexpected info: {'EXIT',<0.581.0>,shutdown}
```

Any idea what happens?

I'm using ejabberd 18.09 but I see no changes to `mod_http_upload_quota.erl` on more recent versions.

**Comment by licaon-kter** [comment]
I have the same 30, no issues.

What else made you think that this is the issue?

**Comment by joudinet** [comment]
Do you also have the same error message on the log?

Have you also the `rm_on_unregister: false` option on `mod_http_upload`?

To be honest, I'm not sure it is the reason of the issue. I just know it is the last modification I've done.
As soon as someone told me she couldn't connect from her chat client, I log in on the server and find out ejabberd was not running anymore. The last line of `ejabberd.log` was this error message.

Since then, I've restarted ejabberd and it is still up after one hour. I'll tell you if ejabberd stops again.

**Comment by licaon-kter** [comment]
What does `error.log` say?

No, I don't have the other option though.

**Comment by joudinet** [comment]
The `error.log` content:
```
# cat /var/log/ejabberd/error.log 
2019-03-04 11:27:24.868 [error] <0.580.0>@mod_http_upload_quota:handle_info:214 Got unexpected info: {'EXIT',<0.581.0>,shutdown}
```

What do you mean by letting ejabberd handle http_upload?
I do have an nginx server that proxies the XEP-0363 requests to ejabberd. So, I would answer yes, I let ejabberd handle the http_upload service, even though I have an nginx proxy. Here the corresponding nginx configuration, in case it is relevant:
```
# Forward XEP-0363 requests to the XMPP server's HTTP listener.
location ^~ /upload {
        proxy_pass       https://127.0.0.1:5443/upload;
        proxy_set_header Host           $host;
        proxy_set_header X-Real-IP      $remote_addr;
        proxy_buffering  off;
        tcp_nodelay      on;
        client_max_body_size 0;
}
```

Let me rephrase my issue: without activating mod_http_upload_quota, the http upload service works well. As soon as I activate mod_http_upload_quota to delete files after 30 days, ejabberd works for a few hours then stops and the only error in the error.log is the one related to `mod_upload_quota:handle_info`

**Comment by licaon-kter** [comment]
External app, nginx but not proxy say via https://github.com/weiss/ngx_http_upload

**Comment by joudinet** [comment]
I see. So no, I'm not using an external app to handle http upload.
I'm using mod_http_upload configured as a request_handler of ejabberd_http module:
```
listen:
  -
    port: 5222
    ip: "::"
    module: ejabberd_c2s
    max_stanza_size: 262144
    shaper: c2s_shaper
    access: c2s
    starttls: true
  -
    port: 5269
    ip: "::"
    module: ejabberd_s2s_in
    max_stanza_size: 524288
  -
    port: 5443
    ip: "::"
    module: ejabberd_http
    request_handlers:
      "/api": mod_http_api
      "/bosh": mod_bosh
      "/upload": mod_http_upload
      "/ws": ejabberd_http_ws
    web_admin: true
    captcha: true
    tls: true
```
Does mod_http_upload_quota requires to use an external app for http upload? When I read the documentation, I understand on the contrary, that mod_http_upload_quota extends mod_http_upload:
`This module adds quota support for mod_http_upload.`

**Comment by licaon-kter** [comment]
> Does mod_http_upload_quota requires to use an external app for http upload?

No, just asking to cover all cases.

I've added that line to a test machine, and first try it couldn't delete the files as I forgot to set the right folder perms, but no crash.

Now I've fixed perms so I'll wait for the next cycle.

**Comment by weiss** [comment]
I'm not happy with the `[error]` message and will look into it, but from what you wrote we can't tell the reason for ejabberd shutting down.  If there's no hint in your `ejabberd.log`, I'd check the `dmesg` and `journalctl`/syslog output.

**Comment by joudinet** [comment]
Unfortunately, there is nothing more in `ejabberd.log`. Here the last 3 lines:
```
2019-03-01 17:56:20.111 [info] <0.1902.0>@ejabberd_c2s:process_terminated:262 (websocket|<0.1901.0>) Closing c2s session for 43481068685385632141443@gnyoi.com/69403570126715694691459: Connection failed: connection closed
2019-03-01 17:56:20.206 [info] <0.1902.0>@mod_push:remove_user:420 Removing any push sessions of 43481068685385632141443@gnyoi.com
2019-03-04 11:27:24.868 [error] <0.580.0>@mod_http_upload_quota:handle_info:214 Got unexpected info: {'EXIT',<0.581.0>,shutdown}
```
There is no message related to ejabberd in dmesg and systemd is not running in this server.

Loglevel is set to 4. I'm going to set it to 5 and restart ejabberd to see if debug messages could help.

**Comment by licaon-kter** [comment]
I got `[debug] <0.416.0>@mod_http_upload_quota:handle_info:196 Got 'sweep' message for mydomain.tld` after a while, but no crash or issue.

@weiss ^^^^ that's the correct message showing that a clean up pass was tried?

**Comment by weiss** [comment]
@licaon-kter 
> that's the correct message showing that a clean up pass was tried?

Yes.

And I doubt that `mod_http_upload_quota` somehow manages to shut down ejabberd (except  maybe if AppArmor/SELinux/whatever goes mad on the filesystem access or something).  My guess would be that ejabberd is shut down for some unrelated reason and `mod_http_upload_quota` just logs this ugly `[error]` during shutdown.

**Comment by zinid** [comment]
This also might be an Erlang bug. The behavior is very weird. What OS and Erlang version are you using?
BTW, is there anything in the crash.log?

**Comment by zinid** [comment]
Also, are you sure "ejabberd stops with the following error"? Did you compare the shutdown time with the time from this error message?

**Comment by zinid** [comment]
Also, run the following commands to find a possible error reason:
```
# zgrep beam.smp /var/log/messages*
# find / -name erl_crash.dump
```
The second command will take some time as it will search the whole disk.

**Comment by joudinet** [comment]
No, I'm not sure "ejabberd stops with the following error", but I'm pretty confident it stops at the same time. Since I've restarted ejabberd with debug log, Yesterday, it hasn't stopped, and I forgot to check `crash.log` before rebooting :-(
I'm using Erlang 21 on Buildroot 2018.08.2.
I agree with @weiss's guess that ejabberd stops for an unrelated reason and it is just mod_http_upload_quota that logs this unexpected shutdown as an error.

Actually, I may have an hypothesis for the responsible of ejabberd shutdown, even though it is really weird. The time of the error message coincides to when of my colleague tested to connect, via openvpn tunnels, the network on which this ejabberd server runs to another network that has another ejabberd server with the same .erlang_cookie. The openvpn was wrongly configured (i.e., compression enabled with a buggy compression algorithm: --compress lz4). Thus, packets going through this tunnel could have their first byte changed to 0xfb.
No idea why this may lead to an ejabberd shutdown but I'll try to reactivate these tunnels to see if I can reproduce ejabberd shutdown.

**Comment by joudinet** [comment]
I managed to reproduce ejabberd shutdown. I have a cluster of 4 ejabberd instances. Few minutes after activating the buggy openvpn tunnels, two of ejabberd nodes went down, the two remaining nodes are still running so far. Debug info is enabled on all of them.
Here the last lines of `ejabberd.log` from the two ejabberd servers that are down:
```
[root@testbed-1 default]# tail /var/log/ejabberd/ejabberd.log 
2019-03-05 11:02:06.333 [debug] <0.1150.0>@ejabberd_access_permissions:handle_call:139 Command 'list_cluster' execution allowed by rule 'console commands' (CallerInfo=#{caller_module => ejabberd_ctl})
2019-03-05 11:02:06.333 [debug] <0.4583.0>@ejabberd_commands:do_execute_command:494 Executing command ejabberd_admin:list_cluster with Args=[]
2019-03-05 11:02:08.323 [debug] <0.1150.0>@ejabberd_access_permissions:handle_call:139 Command 'connected_users_number' execution allowed by rule 'console commands' (CallerInfo=#{caller_module => ejabberd_ctl})
2019-03-05 11:02:08.324 [debug] <0.4586.0>@ejabberd_commands:do_execute_command:494 Executing command ejabberd_sm:connected_users_number with Args=[]
2019-03-05 11:06:00.212 [debug] <0.1601.0>@mod_fail2ban:handle_info:134 cleaning failed_auth ETS table
2019-03-05 11:16:00.214 [debug] <0.1601.0>@mod_fail2ban:handle_info:134 cleaning failed_auth ETS table
2019-03-05 11:26:00.216 [debug] <0.1601.0>@mod_fail2ban:handle_info:134 cleaning failed_auth ETS table
2019-03-05 11:33:48.753 [error] <0.1575.0>@mod_http_upload_quota:handle_info:214 Got unexpected info: {'EXIT',<0.582.0>,shutdown}
2019-03-05 11:33:48.760 [debug] <0.1010.0>@ejabberd_mnesia:schema:203 No custom Mnesia schema for table 'session' found
2019-03-05 11:33:48.761 [info] <0.1010.0>@ejabberd_mnesia:change_table_copy_type:162 Changing Mnesia table 'session' from unknown to ram_copies

[root@testbed-2 default]# tail /var/log/ejabberd/ejabberd.log 
2019-03-05 10:59:14.259 [debug] <0.5274.0>@ejabberd_commands:do_execute_command:494 Executing command ejabberd_sm:connected_users_number with Args=[]
2019-03-05 11:00:49.972 [debug] <0.3355.0>@mod_fail2ban:handle_info:134 cleaning failed_auth ETS table
2019-03-05 11:02:06.574 [debug] <0.3204.0>@ejabberd_access_permissions:handle_call:139 Command 'list_cluster' execution allowed by rule 'console commands' (CallerInfo=#{caller_module => ejabberd_ctl})
2019-03-05 11:02:06.574 [debug] <0.5283.0>@ejabberd_commands:do_execute_command:494 Executing command ejabberd_admin:list_cluster with Args=[]
2019-03-05 11:02:08.586 [debug] <0.3204.0>@ejabberd_access_permissions:handle_call:139 Command 'connected_users_number' execution allowed by rule 'console commands' (CallerInfo=#{caller_module => ejabberd_ctl})
2019-03-05 11:02:08.587 [debug] <0.5286.0>@ejabberd_commands:do_execute_command:494 Executing command ejabberd_sm:connected_users_number with Args=[]
2019-03-05 11:10:49.974 [debug] <0.3355.0>@mod_fail2ban:handle_info:134 cleaning failed_auth ETS table
2019-03-05 11:20:49.977 [debug] <0.3355.0>@mod_fail2ban:handle_info:134 cleaning failed_auth ETS table
2019-03-05 11:30:49.979 [debug] <0.3355.0>@mod_fail2ban:handle_info:134 cleaning failed_auth ETS table
2019-03-05 11:33:48.777 [error] <0.3350.0>@mod_http_upload_quota:handle_info:214 Got unexpected info: {'EXIT',<0.583.0>,shutdown}
```

Both `crash.log` files are empty.
No `beam.smp` occurrence in /var/log/messages*
But I have on both down ejabberd servers a file in /var/log/ejabberd I haven't seen before:
`/var/log/ejabberd/MnesiaCore.nonode@nohost_1551_782028_863205`
This file has been created at 11:33. So at the same time when `mod_http_upload_quota` reports an unexpected shutdown. I'm not sure how to read this file, but opening it with a text editor, it looks like a crash info report. Here the beginning of the file:
```
~Cl^@^@^@^Xh^Bd^@       crashinfoh^Bk^@!~p got unexpected info: ~tp; ~p~nl^@^@^@^Cd^@^Nmnesia_monitorh^Cd^@^DEXITgd^@^Mnonode@nohost^@^@^@�^@^@^@^@^@d^@^Hshutdownk^@+Hint: check that the disk still is writablejh^Bd^@^Dtimeh^Bh^Cb^@^@^G�a^Ca^Eh^Ca^Ka!a0h^Bd^@^Dselfl^@^@^@^Rh^Bd^@^Rcurrent_stacktracel^@^@^@^Fh^Dd^@
mnesia_libd^@^Mproc_dbg_infoa^Al^@^@^@^Bh^Bd^@^Dfilek^@^Nmnesia_lib.erlh^Bd^@^Dlineb^@^@^C^^jh^Dd^@
mnesia_libd^@^Fmkcorea^Al^@^@^@^Bh^Bd^@^Dfilek^@^Nmnesia_lib.erlh^Bd^@^Dlineb^@^@^B�jh^Dd^@
mnesia_libd^@^Efatala^Bl^@^@^@^Bh^Bd^@^Dfilek^@^Nmnesia_lib.erlh^Bd^@^Dlineb^@^@^C�jh^Dd^@
gen_serverd^@^Ltry_dispatcha^Dl^@^@^@^Bh^Bd^@^Dfilek^@^Ngen_server.erlh^Bd^@^Dlineb^@^@^B}jh^Dd^@
gen_serverd^@
handle_msga^Fl^@^@^@^Bh^Bd^@^Dfilek^@^Ngen_server.erlh^Bd^@^Dlineb^@^@^B�jh^Dd^@^Hproc_libd^@^Oinit_p_do_applya^Cl^@^@^@^Bh^Bd^@^Dfilek^@^Lproc_lib.erlh^Bd^@^Dlinea�jjh^Bd^@^Oregistered_named^@^Nmnesia_monitorh^Bd^@^Pcurrent_functionh^Cd^@
mnesia_libd^@^Mproc_dbg_infoa^Ah^Bd^@^Linitial_callh^Cd^@^Hproc_libd^@^Finit_pa^Eh^Bd^@^Fstatusd^@^Grunningh^Bd^@^Qmessage_queue_lena^@h^Bd^@^Elinksl^@^@^@^Bgd^@^Mnonode@nohost^@^@^C�^@^@^@^@^@gd^@^Mnonode@nohost^@^@^D^T^@^@^@^@^@jh^Bd^@
dictionaryl^@^@^@^Bh^Bd^@^M$initial_callh^Cd^@^Nmnesia_monitord^@^Dinita^Ah^Bd^@
$ancestorsl^@^@^@^Cd^@^Qmnesia_kernel_supd^@
mnesia_supgd^@^Mnonode@nohost^@^@^C�^@^@^@^@^@jjh^Bd^@  trap_exitd^@^Dtrueh^Bd^@^Merror_handlerd^@^Merror_handlerh^Bd^@^Hpriorityd^@^Fnormalh^Bd^@^Lgroup_leadergd^@^Mnonode@nohost^@^@^C�^@^@^@^@^@h^Bd^@^Ototal_heap_sizeb^@^@^G�h^Bd^@       heap_sizeb^@^@^C�h^Bd^@
stack_sizea3h^Bd^@
```

And I have also a file ` erl_crash_20190304-152601.dump` file on testbed-2, created today at 11:34. Here the beginning of this file:
```
=erl_crash_dump:0.5
Tue Mar  5 11:34:00 2019
Slogan: Kernel pid terminated (application_controller) ({application_terminated,kernel,shutdown})
System version: Erlang/OTP 21 [erts-10.0] [source] [smp:2:2] [ds:2:2:10] [async-threads:1]
Compiled: Tue Feb 26 20:46:37 2019
Taints: asn1rt_nif,ets_cache,xmpp_lang,xmpp_uri,jid,ezlib_drv,stringprep,fxml_stream,fxml,fast_tls,p1_sha,fast_yaml,crypto
Atoms: 29547
Calling Thread: scheduler:1
=scheduler:1
Scheduler Sleep Info Flags:
Scheduler Sleep Info Aux Work: THR_PRGR_LATER_OP
Current Port:
Run Queue Max Length: 0
Run Queue High Length: 0
Run Queue Normal Length: 0
Run Queue Low Length: 0
Run Queue Port Length: 0
Run Queue Flags: OUT_OF_WORK | HALFTIME_OUT_OF_WORK | NONEMPTY | EXEC
Current Process: <0.0.0>
Current Process State: Running
Current Process Internal State: ACT_PRIO_NORMAL | USR_PRIO_NORMAL | PRQ_PRIO_NORMAL | ACTIVE | RUNNING
Current Process Program counter: 0xb5d24b20 (init:sleep/1 + 40)
Current Process CP: 0x00000000 (invalid)
Current Process Limited Stack Trace:
0xb32079e0:SReturn addr 0xB5D23348 (init:loop/1 + 320)
0xb32079e4:SReturn addr 0x23D240 (<terminate process normally>)
=scheduler:2
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Port:
Run Queue Max Length: 0
Run Queue High Length: 0
Run Queue Normal Length: 0
Run Queue Low Length: 0
Run Queue Port Length: 0
Run Queue Flags: OUT_OF_WORK | HALFTIME_OUT_OF_WORK
Current Process:
=dirty_cpu_scheduler:3
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_cpu_scheduler:4
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_cpu_run_queue
Run Queue Max Length: 0
Run Queue High Length: 0
Run Queue Normal Length: 0
Run Queue Low Length: 0
Run Queue Port Length: 0
Run Queue Flags: OUT_OF_WORK | HALFTIME_OUT_OF_WORK
=dirty_io_scheduler:5
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_io_scheduler:6
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_io_scheduler:7
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_io_scheduler:8
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
```

**Comment by joudinet** [comment]
FYI, I managed to read the mnesia core dump:
```
[{crashinfo,
     {"~p got unexpected info: ~tp; ~p~n",
      [mnesia_monitor,
       {'EXIT',<0.250.0>,shutdown},
       "Hint: check that the disk still is writable"]}},
 {time,{{2019,3,5},{11,33,48}}},
 {self,
     [{current_stacktrace,
          [{mnesia_lib,proc_dbg_info,1,
               [{file,"mnesia_lib.erl"},{line,798}]},
           {mnesia_lib,mkcore,1,[{file,"mnesia_lib.erl"},{line,682}]},
           {mnesia_lib,fatal,2,[{file,"mnesia_lib.erl"},{line,967}]},
           {gen_server,try_dispatch,4,
               [{file,"gen_server.erl"},{line,637}]},
           {gen_server,handle_msg,6,
               [{file,"gen_server.erl"},{line,711}]},
           {proc_lib,init_p_do_apply,3,
               [{file,"proc_lib.erl"},{line,249}]}]},
      {registered_name,mnesia_monitor},
      {current_function,{mnesia_lib,proc_dbg_info,1}},
      {initial_call,{proc_lib,init_p,5}},
      {status,running},
      {message_queue_len,0},
      {links,[<0.2816.0>,<0.2837.0>]},
      {dictionary,
          [{'$ancestors',[mnesia_kernel_sup,mnesia_sup,<0.2812.0>]},
           {'$initial_call',{mnesia_monitor,init,1}}]},
      {trap_exit,true},
      {error_handler,error_handler},
      {priority,normal},
      {group_leader,<0.2811.0>},
      {total_heap_size,4184},
      {heap_size,2586},
      {stack_size,51},
      {reductions,55724},
      {garbage_collection,
          [{max_heap_size,#{error_logger => true,kill => true,...}},
           {min_bin_vheap_size,46422},
           {min_heap_size,233},
           {fullsweep_after,...},
           {...}]},
      {suspending,[]}]},
 {nodes,{[{nonode@nohost,<0.5347.0>}],[]}},
```

**Comment by joudinet** [comment]
I've got again these mnesia crashes with an openvpn without compression. So, I have to rule out the bug in the compression algorithm in openvpn as the explanation :-/
So, my current settings is a cluster of 4 ejabberd working correctly. Then, I open an openvpn tap tunnel to join another 3 servers running ejabberd as well, with the same `.erlang_cookie`. Few minutes later, all 4 ejabberd are down with no message in the log except the error from mod_http_upload_quota, an Mnesia core file and an erl crash dump.
Has anyone try to run a cluster of ejabberd servers connected via openvpn tap tunnels before?

**Comment by zinid** [comment]
I honestly do not know how to help you with the issue since reproducing it requires lots of setup on our side.

**Comment by joudinet** [comment]
I understand. Actually we have no deployment with such architecture yet. This was only a testbed to test a scenario for a client request. So it is hard for me as well to reproduce it. Maybe just keep in mind that they may have mnesia and erl crashes when ejabberd run through OpenVPN tunnels?
I don't know if it is possible, but maybe have a way to log such crashes in ejabberd's log files?

**Comment by zinid** [comment]
> maybe have a way to log such crashes in ejabberd's log files?

No, Mnesia cores are not interceptable.

**Comment by zinid** [comment]
I close the issue. It may be reopened when more info is provided.
